test/integration-test/src/tests/relocations.rs

```diff
@@ -0,0 +1,247 @@
+    fn build_ebpf(&self) -> Result<Vec<u8>> {
+        let local_definition = self.local_definition;
+        let relocation_code = self.relocation_code;
+        let tmp_dir = tempfile::tempdir().context("Error making temp dir")?;
+        let ebpf_file = tmp_dir.path().join("ebpf_program.c");
+        std::fs::write(
+            &ebpf_file,
+            format!(
+                r#"
+                    #include <linux/bpf.h>
+
+                    #include <bpf/bpf_core_read.h>
+                    #include <bpf/bpf_helpers.h>
+                    #include <bpf/bpf_tracing.h>
+
+                    {local_definition}
+
+                    struct {{
+                      __uint(type, BPF_MAP_TYPE_ARRAY);
+                      __type(key, __u32);
+                      __type(value, __u32);
+                      __uint(max_entries, 1);
+                    }} output_map SEC(".maps");
+
+                    SEC("tracepoint/bpf_prog") int bpf_prog(void *ctx) {{
+                      __u32 key = 0;
+                      {relocation_code}
+                      bpf_map_update_elem(&output_map, &key, &value, BPF_ANY);
+                      return 0;
+                    }}
+
+                    char _license[] SEC("license") = "GPL";
+                "#
+            ),
+        )
+        .context("Writing bpf program failed")?;
+        Command::new("clang")
+            .current_dir(tmp_dir.path())
+            .args(["-c", "-g", "-O2", "-target", "bpf"])
+            // NOTE: these tests depend on libbpf, LIBBPF_INCLUDE must point its headers.
+            // This is set automatically by the integration-test xtask.
+            .args([
+                "-I",
+                &std::env::var("LIBBPF_INCLUDE").context("LIBBPF_INCLUDE not set")?,
+            ])
+            .arg(&ebpf_file)
+            .status()
+            .context("Failed to run clang")?
+            .success()
+            .then_some(())
+            .context("Failed to compile eBPF program")?;
+        let ebpf = std::fs::read(ebpf_file.with_extension("o"))
+            .context("Error reading compiled eBPF program")?;
+        Ok(ebpf)
+    }
+
+    /// - Generate the target BTF source with a mock main()
+    /// - Compile it with clang
+    /// - Extract the BTF with pahole
+    fn build_btf(&self) -> Result<Btf> {
+        let target_btf = self.target_btf;
+        let tmp_dir = tempfile::tempdir().context("Error making temp dir")?;
+        // BTF files can be generated and inspected with these commands:
+        // $ clang -c -g -O2 -target bpf target.c
+        // $ pahole --btf_encode_detached=target.btf -V target.o
+        // $ bpftool btf dump file ./target.btf  format c
+        let btf_file = tmp_dir.path().join("target_btf.c");
+        std::fs::write(
+            &btf_file,
+            format!(
+                r#"
+                    #include <linux/types.h>
+                    {target_btf}
+                    int main() {{ return 0; }}
```

Comment:
As this is now, the compiler isn't guaranteed to emit BTF info for the target btf, since it's unused and could be dead-code eliminated. In fact this happens with the pointer test (once you add the target declarations, which are now missing). 

One workaround is to compile the tracepoint program even for the target definition. Otherwise we need to ensure that all target definitions are actually definitions (they create globals or something), instead of just being declarations.

Reply:
Interesting, thanks!